File: przygotowanie_pliki_liczby/c++/liczby/liczby.cpp
// Z funckjami wbudowanymi

#include <iostream>
#include <string>
#include <bitset>
#include <sstream>

void dec_to_bin(int liczba) {
    // Zamienia liczbę dziesiętną na liczbę dwójkową (binarną).
    std::cout << std::bitset<32>(liczba) << std::endl; // 32-bitowa reprezentacja binarna
}

void dec_to_oct(int liczba) {
    // Zamienia liczbę dziesiętną na liczbę ósemkową (oktalną).
    std::cout << std::oct << liczba << std::dec << std::endl; // Wyświetl jako ósemkową
}

void dec_to_hex(int liczba) {
    // Zamienia liczbę dziesiętną na liczbę szesnastkową (heksadecymalną).
    std::cout << std::hex << liczba << std::dec << std::endl; // Wyświetl jako heksadecymalną
}

void bin_to_dec(const std::string& liczba_bin) {
    // Zamienia liczbę dwójkową na liczbę dziesiętną.
    std::cout << std::stoi(liczba_bin, nullptr, 2) << std::endl;
}

void oct_to_dec(const std::string& liczba_oct) {
    // Zamienia liczbę ósemkową na liczbę dziesiętną.
    std::cout << std::stoi(liczba_oct, nullptr, 8) << std::endl;
}

void hex_to_dec(const std::string& liczba_hex) {
    // Zamienia liczbę szesnastkową na liczbę dziesiętną.
    std::cout << std::stoi(liczba_hex, nullptr, 16) << std::endl;
}

void x_to_dec(const std::string& liczba_x, int system) {
    // Zamienia liczbę z dowolnego systemu na liczbę dziesiętną.
    std::cout << std::stoi(liczba_x, nullptr, system) << std::endl;
}

int main() {
    // Deklaracja main =/= uruchomienie go

    // Liczby dziesiętne
    std::cout << "Podaj liczbę dziesiętną: ";
    int inp;
    std::cin >> inp;

    dec_to_bin(inp);
    dec_to_oct(inp);
    dec_to_hex(inp);

    // Liczba binarna
    std::cout << "Podaj liczbę binarną (np. 10101): ";
    std::string inp_bin;
    std::cin >> inp_bin;
    bin_to_dec(inp_bin);

    // Liczba oktalna
    std::cout << "Podaj liczbę octalną (np. 17171): ";
    std::string inp_oct;
    std::cin >> inp_oct;
    oct_to_dec(inp_oct);

    // Liczba heksadecymalna
    std::cout << "Podaj liczbę heksadecymalną (np. ABCDF): ";
    std::string inp_hex;
    std::cin >> inp_hex;
    hex_to_dec(inp_hex);

    // Liczba w dowolnym systemie
    std::cout << "Podaj liczbę w dowolnym systemie: ";
    std::string inp_x;
    std::cin >> inp_x;
    std::cout << "Podaj system liczbowy (np. 3 to trójkowy): ";
    int base;
    std::cin >> base;
    x_to_dec(inp_x, base);

    return 0;
}

File: przygotowanie_pliki_liczby/python/liczby/liczby.py
"""
 Przykłady z liczbami przy użyciu funckji wbudowanych.

 Każdy przykład ma osobne funkcje.

 Kod jest też w miarę samodokumentujący i jest napisany zgodnie z formatowaniem python.
 niektórych rzeczy takich jak ': int' czy '-> int' nie trzeba pisać ale kij.
"""

def dec_to_bin(liczba: int) -> None:
    """
     Zamienia liczbę dziesiętną na liczbę dwójkową (binarną).
    """

    base_2: str = bin(liczba)
    print(base_2)  # "0bxxxx"
    # Możemy też obciąć "0b"
    print(base_2[2::])

    """
    [x:y:z] oznacza: x - indeks startowy, y: indeks końcowy, z: krok
    [2::] zacznij od drugiego elementu (czyli [1]) i idź do końca
    [::-1] zacznij od zerowego elementu i idź wstecz (czyli od końca do początku)
    """


def dec_to_oct(liczba: int) -> None:
    """
     Zamienia liczbę dziesiętną na liczbę ósemkową (oktalną).
    """

    print(oct(liczba)[2::])


def dec_to_hex(liczba: int) -> None:
    """
     Zamienia liczbę dziesiętną na liczbę szesnastkową (heksadecymalną).
    """

    print(hex(liczba))  # Nie obcinam 0x bo czasem ciężko ogarnąć że to hex.


def bin_to_dec(liczba_bin: str) -> None:
    """
     Zamienia liczbę dwójkową na liczbę dziesiętną.
    """

    print(int(liczba_bin,2))


def oct_to_dec(liczba_oct: str) -> None:
    """
     Zamienia liczbę ósemkową na liczbę dziesiętną.
    """

    print(int(liczba_oct,8))


def hex_to_dec(liczba_hex: str) -> None:
    """
     Zamienia liczbę szesnastkową na liczbę dziesiętną.
    """

    print(int(liczba_hex,16))


def x_to_dec(liczba_x: str, system: int) -> None:
    """
     Zamienia liczbę z dowolnego systemu na liczbę dziesiętną.
    """

    print(int(liczba_x,system))


def main() -> None:
    """
     Deklaracja main =/= uruchomienie go
    """

    # Liczby dziesiętne
    print("Podaj liczbę dziesiętną: ")
    inp: int = int(input())

    dec_to_bin(inp)
    dec_to_oct(inp)
    dec_to_hex(inp)

    # Liczba binarna
    print("Podaj liczbę binarną (np. 10101)")
    inp: str = input()
    bin_to_dec(inp)

    # Liczba oktalna
    print("Podaj liczbę octalną (np. 17171)")
    inp = input()
    oct_to_dec(inp)

    # Liczba heksadecymalna
    print("Podaj liczbę heksadecymalną (np. ABCDF)")
    inp = input()
    hex_to_dec(inp)

    # Liczba w dowolnym systemie
    print("Podaj liczbę w dowolnym systemie")
    inp = input()
    print("Podaj system liczbowy (np. 3 to trójkowy) (czyli najwyższa cyfra/litera + 1)")
    base = int(input())
    x_to_dec(inp, base)



"""
 Nie polecam zostawiać kodu luzem poza funkcjami, zwłaszcza w większych projektach
"""
if __name__ == "__main__":
    main()
